libraries/eosiolib/contracts/eosio/call.hpp

```diff
@@ -46,65 +51,107 @@ namespace eosio {
-
-      /**
-       * Construct a new call object with receiver, name, and payload data
-       *
-       * @tparam T  - Type of call data, must be serializable by `pack(...)`
-       * @param receiver -  The name of the account this call is intended for
-       * @param flags - The flags
-       * @param payload - The call data that will be serialized via pack into data
-       */
-      template<typename T>
-      call( struct name receiver, T&& payload, bool read_only = false, bool no_op = false )
-      : receiver(receiver)
-      , read_only(read_only)
-      , no_op_if_receiver_not_support_sync_call(no_op)
-      , data(pack(std::forward<T>(payload))) {}
-
-      /// @cond INTERNAL
-      EOSLIB_SERIALIZE( call, (receiver)(read_only)(no_op_if_receiver_not_support_sync_call)(data) )
-      /// @endcond
-
-      /**
-       * Make a call using the functor operator
-       */
-      int64_t operator()() const {
-         uint64_t flags = read_only ? 0x01 : 0x00; // last bit indicating read only
-         auto retval =  internal_use_do_not_use::call(receiver.value, flags, data.data(), data.size());
-
-         if (retval == -1) {  // sync call is not supported by the receiver contract
-            check(no_op_if_receiver_not_support_sync_call, "receiver does not support sync call but no_op_if_receiver_not_support_sync_call flag is not set");
+   template <eosio::name::raw Func_Name, auto Func_Ref, access_mode Access_Mode=access_mode::read_write, support_mode Support_Mode = support_mode::abort>
+   struct call_wrapper {
+      template <typename Receiver>
+      constexpr call_wrapper(Receiver&& receiver)
+         : receiver(std::forward<Receiver>(receiver))
+      {}
+
+      static constexpr eosio::name function_name = eosio::name(Func_Name);
+      eosio::name receiver {};
+
+      using orig_ret_type = typename detail::function_traits<decltype(Func_Ref)>::return_type;
+
+      using return_type = std::conditional_t<
+         Support_Mode == support_mode::abort,   // if Support_Mode is abort
+         orig_ret_type,                         // use the original return type
+         std::conditional_t<                    // else
+            std::is_void<orig_ret_type>::value, // original return type is void
+            std::optional<void_call>,           // use optional of empty struct
+            std::optional<orig_ret_type>        // use optional of original return type
+         >
+      >;
+
+      template <typename... Args>
+      return_type operator()(Args&&... args)const {
+         static_assert(detail::type_check<Func_Ref, Args...>());
+
+         uint64_t flags = 0x00;
+         if constexpr (Access_Mode == access_mode::read_only) {
+            flags = 0x01;
+         }
+
+         call_data_header header{ .version   = 0,
+                                  .func_name = function_name.value };
+ 
+         const std::vector<char> data{ pack(std::make_tuple(header, detail::deduced<Func_Ref>{std::forward<Args>(args)...})) };
```

Comment:
`std::make_tuple` is making a copy of the `header` which is no big deal; but it is also making a copy of all the args. I think we should avoid making a copy of the args just to `pack` them. Should be able to use `std::forward_as_tuple` instead of `std::make_tuple`.

Reply:
https://github.com/AntelopeIO/cdt/pull/351/commits/090c99fd66035827cdd73eb91d281f9e13d493f3